app/uk/gov/hmrc/fileupload/controllers/EnvelopeFilesConstraints.scala

```diff
@@ -20,7 +20,8 @@ import scala.util.matching.Regex
 
 case class EnvelopeFilesConstraints(maxItems: Int,
                                     maxSize: Size,
-                                    maxSizePerItem: Size) {
+                                    maxSizePerItem: Size,
+                                    allowZeroLengthFiles: Option[Boolean]) {
```

Comment:
could we just have simple type bool here? With default to none?

Reply:
Here it should be just Boolean with default = true.
Option is valid only at EnvelopeConstraintsUserSetting .

Reply:
Default value is less important but it would make compatibility easy

Reply:
Now on prod. The zero length is accepted. With a flag it will be neither accepted or not. Absence of a flag has no mapping to third state.

Reply:
Or If you like you can mark that giving a flag mark it mean already you got allowZeroLengthFiles = false.
In either case something needs to be changed.

Reply:
There's a complication with making this a Boolean, as it'll no longer be backwards compatible as envelopes currently persisted in mongo do not have this field (this case class maps directly to the mongo collection for persistence).

Making this an Option[Boolean] prevents the need to build / run a backfill / migration script to add this field to existing persisted events.

Could be moved to a `Boolean` later once old events have expired

Reply:
In this change the absence of the flag does default to allowing empty files btw (and does persist the flag)

Reply:
Are you sure it wouldn't work with default value?

Reply:
what about `fobidZeroLenghtFiles: Option[fobidZeroLenghtFiles.type]`?
It would represent the model and would be backwards compatible

Reply:
Ah good point. Yep may well work if we give a default them - wouldn't be able to define the default in configuration then, but  that probably isn't a big deal.
Cheers, will try tomorrow and see if it works

Reply:
After further thought @kastoestoramadus  I'm actually erring back to keeping the `Option[Boolean]` here.
Justification for this is:

* As the `CreateEnvelope` case class is composed of one of these (and `CreateEnvelope` maps directly to a mongo document) `Option[Boolean]` IS an explicit and correct type, as documents created in the past do not have this data - and in new documents it's a `Boolean`

* `noEmptyFilesFlag: Option[SOME_OBEJCT.type]` would hide data, as you couldn't distinguish between envelopes where zero length files are explicitly allowed, and envelopes created before the change.
the option would also be defined in the negative case, which will cause some mind bending and be hard to map back to JSON

* Having this as a `Boolean` with a default value would work but has a couple of drawbacks:
  * You can't default this through configuration (as is done here) as it would no longer be a compile time constant. it could well be desirable to default to disallowing empty files in the future
  * we don't have integration tests that cover state from previous mongo schema migrations. they would be good to have, but probably out of scope given the context of the project now.
    this is important as the default could be innocently removed and then inlined, tests would pass (as it's OK to do that for new envelopes) but calls to retrieve envelope data created in a previous FUAAS version would then fail at runtime.
    this is covered currently in this proposed change 

Would be good to hear your feedback on this justification, just thinking `Option[Boolean]` is the best compromise ATM until test coverage of schema migrations is possible (time being the main blocker in that).

Reply:
ad. third point. I doubt the need to change default value. Even if it would be possible then such change of default behaviour would break the behaviour of older envelopes.
In both cases you need an additional data migration effort.

ad. second point.  "you couldn't distinguish between envelopes where zero length files are explicitly allowed" <- this is not true, I've proposed a negated flag to yours. Absense of flag mean the same before and with new flag feature.

Reply:
Don't think it would break behaviour to change the default value? Would just change behaviour of new envelopes created without an explicit value for the flag :- previously created envelopes would have the same value

Not sure I follow that second point sorry:
* in envelopes created before this change, the option would not be defined
* in envelopes where empty files are explicitly allowed, the option would not be defined
my point was that the user of this flag (frontend) can't distinguish between these, and to have the flag present in the negative case will bring a need for overly complex branches elsewhere

Given the context of the project and the expediency to do this change I'll merge unless you think it's fundamentally broken